core/src/test/java/org/apache/iceberg/TestOverwrite.java

```diff
@@ -312,4 +313,154 @@ public void testValidatedOverwriteWithAppendSuccess() {
+  }
+
+  @Test
+  public void testOverwriteWithDeleteFileAndAppendManifest() throws IOException {
+    // merge all manifests for this test
+    super.table.updateProperties().set("commit.manifest.min-count-to-merge", "1").commit();
+
+    Assert.assertNull("Should not have a current snapshot", latestSnapshot(readMetadata(), branch));
+    Assert.assertEquals("Table should start empty", 0, listManifestFiles().size());
+
+    commit(
+        super.table,
+        super.table.newOverwrite().addFile(FILE_C).addFile(FILE_D),
+        branch); // Added data file , FILE_C and FILE_D
+
+    Assert.assertNotNull("Should create a snapshot", latestSnapshot(super.table, branch));
+    V1Assert.assertEquals(
+        "Last sequence number should be 0", 0, super.table.ops().current().lastSequenceNumber());
+    V2Assert.assertEquals(
+        "Last sequence number should be 1", 1, super.table.ops().current().lastSequenceNumber());
+
+    Snapshot commitBefore = latestSnapshot(super.table, branch);
+    long baseId = commitBefore.snapshotId();
+    validateSnapshot(null, commitBefore, 1, FILE_C, FILE_D);
+
+    Assert.assertEquals(
+        "Should create 1 manifest for initial write",
+        1,
+        commitBefore.allManifests(table.io()).size());
+    ManifestFile initialManifest = commitBefore.allManifests(table.io()).get(0);
+    validateManifest(
+        initialManifest,
+        dataSeqs(1L, 1L),
+        fileSeqs(1L, 1L),
+        ids(baseId, baseId),
+        files(FILE_C, FILE_D),
+        statuses(Status.ADDED, Status.ADDED));
+
+    ManifestFile manifest = writeManifest(FILE_A, FILE_B); // Added Data File FILE_A, FILE_B
+    commit(
+        super.table,
+        super.table
+            .newOverwrite()
+            .appendManifest(manifest)
+            .deleteFile(FILE_C), // Delete Data File FILE_C
+        branch);
+
+    V1Assert.assertEquals(
+        "Last sequence number should be 0", 0, super.table.ops().current().lastSequenceNumber());
+    V2Assert.assertEquals(
+        "Last sequence number should be 2", 2, super.table.ops().current().lastSequenceNumber());
+
+    Snapshot committedAfter = latestSnapshot(super.table, branch);
+    Assert.assertEquals(
+        "Should contain 1 merged manifest for second write",
+        1,
+        committedAfter.allManifests(super.table.io()).size());
+    ManifestFile newManifest = committedAfter.allManifests(super.table.io()).get(0);
+    Assert.assertNotEquals(
+        "Should not contain manifest from initial write", initialManifest, newManifest);
+
+    long snapshotId = committedAfter.snapshotId();
+
+    validateManifest(
+        newManifest,
+        dataSeqs(2L, 2L, 1L, 1L),
+        fileSeqs(2L, 2L, 1L, 1L),
+        ids(snapshotId, snapshotId, snapshotId, baseId),
+        concat(files(FILE_A, FILE_B), files(initialManifest)), // FILE_A, FILE_B, FILE_C, FILE_D
+        statuses(Status.ADDED, Status.ADDED, Status.DELETED, Status.EXISTING)); // FILE_C - DELETED
+  }
```

Comment:
Can we add more testcase which throws exception when delete manifest is added etc?